Add OLM mirroring using offline media (#30230)
openshift/openshift-docs: modules/olm-mirroring-catalog.adoc

----
endif::[]

> . The `oc adm catalog mirror` command extracts the contents of an index image to generate the manifests required for mirroring. The default behavior of the command generates manifests, then automatically mirrors all of the image content from the index image, as well as the index image itself, to your mirror registry. Alternatively, if your mirror registry is on an airgapped host, you can first mirror the content to removable media, move the media to the disconnected environment, then mirror the content from the media to the registry.

Author: Only meant to mention as a first-term-style hit for search hits, didn't mean to use this much!

** *Option A:* If your mirror registry is on the same network as your workstation with unrestricted network access, run the following command on the workstation:
+
[source,terminal]
----
$ oc adm catalog mirror \
    <index_image> \//<1>
    <mirror_registry>:<port>/<namespace> \//<2>
    [-a ${REG_CREDS}] \//<3>
    [--insecure] \//<4>
    [--index-filter-by-os='<platform>/<arch>'] \//<5>
    [--manifests-only] <6>
----
<1> Specify the index image for the catalog you want to mirror. For example, this might be a pruned index image that you created previously, or one of the source index images for the default catalogs, such as `{index-image-pullspec}`.

> <2> Specify the target registry and namespace to mirror the Operator content to, where `<namespace>` is any existing namespace on the registry. For example, you might create an `olm-mirror` namespace to push all mirrored content to.

Reviewer: If you want users to make a dedicated namespace for the registry before they get here, I'd make it a prereq.

using database path mapping: /database/index.db:/tmp/153048078
wrote database to /tmp/153048078 <1>
...
wrote mirroring manifests to manifests-{index-image}-1614211642 <2>
----
<1> Directory for the temporary `index.db` database generated by the command.
<2> Note the name of the manifests directory that is generated.

** *Option B:* If your mirror registry is on a completely disconnected, or airgapped, host, take the following actions.

.. Run the following command on your workstation with unrestricted network access to mirror the content to local files:
+
[source,terminal]
----
$ oc adm catalog mirror \
    <index_image> \//<1>
    file:///local/index \//<2>
    [-a ${REG_CREDS}] \
    [--insecure]
----
<1> Specify the index image for the catalog you want to mirror. For example, this might be a pruned index image that you created previously, or one of the source index images for the default catalogs, such as `{index-image-pullspec}`.

> <2> Mirrors content to local files in your current directory.

Author: Added mention of `--dir`.

> <2> Mirrors content to local files in your current directory.

Author: Can't actually get `--dir` to work (seems like it's supposed to work like other commands' `--to-dir` per https://github.com/openshift/oc/pull/611, but my destination folder remains empty and it still just copies to the current working directory when I try), so I'm removing mention of `--dir` for now.

+
.Example output
[source,terminal]
----
...
info: Mirroring completed in 5.93s (5.915MB/s)
wrote mirroring manifests to manifests-my-index-1614985528 <1>

To upload local images to a registry, run:

	oc adm catalog mirror file://local/index/myrepo/my-index:v1 REGISTRY/REPOSITORY <2>
----
<1> Note the name of the manifests directory that is generated.
<2> Note the file path given in the example command.

.. Copy the `v2/` directory that is generated in your current directory to removable media.

.. Physically remove the media and attach it to a host in the disconnected environment that has access to the mirror registry.

.. Run the following command on the host in the disconnected environment to upload the images to the mirror registry:
+
[source,terminal]
----
$ oc adm catalog mirror \
    file://local/index/<repo>/<index_image>:<tag> \//<1>
    <mirror_registry>:<port>/<namespace> \//<2>
    [-a ${REG_CREDS}] \
    [--insecure]
----

> <1> Specify the file path that was given in the previous command output.

Reviewer: ```suggestion
<1> Specify the file path from the previous command output.
```

<2> Specify the target registry and namespace to mirror the Operator content to, where `<namespace>` is any existing namespace on the registry. For example, you might create an `olm-mirror` namespace to push all mirrored content to.

> . After mirroring the content to the registry, inspect the manifests directory that is generated in your current directory. If you chose Option A in the previous step, the manifests directory name takes the following form:

Reviewer: I already forgot which one was option A and which one was option B. 😂  I'd probably say something like "If your registry is attached to your network" and "If your registry is in a disconnected environment" instead.

I might also s/inspect/record or add a note at the end that you're going to need this directory name. IDK if I'm just not paying attention today, but when I got to the next step, I did not know that you meant this directory.

Author: I originally had it how you're suggesting, and switched it to referring just to the "Options" to try to shorten it up, but it's clear that is not helpful so will revert!

====
* The `imageContentSourcePolicy.yaml` file defines an `ImageContentSourcePolicy` object that can configure nodes to translate between the image references stored in Operator manifests and the mirrored registry.
* The `mapping.txt` file contains all of the source images and where to map them in the target registry. This file is compatible with the `oc image mirror` command and can be used to further customize the mirroring configuration.
+
[IMPORTANT]
====
If you used the `--manifests-only` flag during the mirroring process and want to further trim the subset of packages to be mirrored, see the steps in the "Mirroring a Package Manifest Format catalog image" procedure about modifying your `mapping.txt` file and using the file with the `oc image mirror` command. After following those further actions, you can continue this procedure.
====
--

. On a host with access to the disconnected cluster, create the `ImageContentSourcePolicy` object by running the following command to specify the `imageContentSourcePolicy.yaml` file in your manifests directory:
+
[source,terminal,subs="attributes+"]
----
$ oc create -f <path/to/manifests/dir>/imageContentSourcePolicy.yaml
----
+

> where `<path/to/manifests/dir>` depends on which option you chose during the mirroring process.

Reviewer: Maybe:
```suggestion
where `<path/to/manifests/dir>` is the path to the manifests directory for your mirrored content.
```
